src/services/drive.ts

```diff
@@ -29,6 +29,195 @@ const mimeLookup = (path: string) => {
+  uid: string,
+  fileToUpload: File
+): Promise<{ id: string }> => {
+  updateBlock({
+    uid,
+    text: `Loading ${Math.round((100 * start) / contentLength)}%`,
+  });
+  const end = Math.min(start + CHUNK_MAX, contentLength);
+  const reader = new FileReader();
+  reader.readAsArrayBuffer(fileToUpload.slice(start, end));
+  return new Promise((resolve, reject) => {
+    reader.onloadend = () => {
+      console.log(
+        `Uploading chunk starting at byte ${start} and ending at byte ${end}`
+      );
+      const buf = new Uint8Array(reader.result as ArrayBuffer);
+      const contentRange = `bytes ${start}-${end - 1}/${contentLength}`;
+      console.log("apiPut", {
+        headers: { contentLength, contentRange },
+      });
+      // Execute the PUT request to upload the chunk
+      fetch(`${domain}${path}`, {
+        method: "PUT",
+        headers: {
+          Authorization: Authorization,
+          "Content-Range": contentRange,
+          "Content-Length": String(end - start),
+        },
+        body: buf,
+      })
+        .then((r) => {
+          if (r.status === 200 || r.status === 201) {
+            // Upload is complete
+            return r.json();
+          } else if (r.status === 308) {
+            // Resume Incomplete, get the next byte range to start uploading from
+            const rangeHeader = r.headers.get("Range");
+            if (rangeHeader) {
+              const nextStart =
+                Number(rangeHeader.replace(/^bytes=0-/, "")) + 1;
+              return uploadFileChunk(
+                nextStart,
+                domain,
+                path,
+                Authorization,
+                contentLength,
+                uid,
+                fileToUpload
+              );
+            } else {
+              // If Range header is missing, start from beginning
+              return uploadFileChunk(
+                0,
+                domain,
+                path,
+                Authorization,
+                contentLength,
+                uid,
+                fileToUpload
+              );
+            }
+          } else if (r.status === 404) {
+            return Promise.reject(
+              "Upload session expired, restart the upload."
+            );
+          } else {
+            console.error(`Server responded with ${r.status}: ${r.statusText}`);
+            return Promise.reject(
+              `Server responded with ${r.status}: ${r.statusText}`
+            );
+          }
+        })
+        .then((data) => {
+          if (data && data.id) {
+            console.log("Upload complete");
+            resolve({ id: data.id });
+          }
+        })
+        .catch((err) => {
+          console.error(`Upload failed: ${err}`);
+          return uploadFileChunk(
+            start,
+            domain,
+            path,
+            Authorization,
+            contentLength,
+            uid,
+            fileToUpload
+          );
+        });
```

Comment:
This is the most important part of the PR, the rest is mostly refactoring into functions.

This replaces `apiPost` with `fetch`